Opentelemetry OCI guide (#1017)
micronaut-projects/micronaut-guides: guides/micronaut-cloud-trace-google/micronaut-cloud-trace-google.adoc

external:micronaut-cloud-trace-base/start.adoc[]
* A https://cloud.google.com/gcp/[Google Cloud Platform (GCP)] account and a GCP project.

common:costs-gcloud.adoc[]

common:gcp.adoc[]

external:micronaut-cloud-trace-base/create-app.adoc[]

=== Tracer Configuration

> Using Micronaut CLI or Launch to create your application, the OpenTelemetry exporter will be added to your `application.yml` configurations.

Reviewer: I'm not completely sure I understand what this is supposed to say. Is the exporter added automatically, no matter what, or is it added automatically only if you use Micronaut CLI or Launch?

```suggestion
If you use Micronaut CLI or Launch to create your application, the OpenTelemetry exporter will automatically be added to your `application.yml` configurations.
```
If this is the case, are there other ways to create an application in which the exporter is NOT automatically added? Or should this read more like the following?

```suggestion
Use Micronaut CLI or Launch to create your application. You will see that the OpenTelemetry exporter is automatically added to your `application.yml` configurations.
```

Reviewer 2: Fixed in https://github.com/micronaut-projects/micronaut-guides/pull/964

resource:application.yml[tag=otel-trace]

== Run the Application

> The application can be deployed to Google Cloud Run, Cloud Function, Cloud Compute or App Engine.

Reviewer: ```suggestion
The application can be deployed to Google Cloud Run, Cloud Function, Cloud Compute, or App Engine.
```

Traces can be sent to Google Cloud Trace outside the Google Cloud.

This allows us to run the application locally and see the traces in Google Cloud Trace.

common:runapp-instructions.adoc[]

> [NOTE]

Reviewer: Should this note include a comment like, "This is normal and will not impact your application performance" or "You can disregard this error message" or "If you get this error message, shut down your machine immediately and perform a standard exorcism ritual before turning it back on"?

Reviewer 2: Fixed in https://github.com/micronaut-projects/micronaut-guides/pull/964

====
You might get this error message: +
"Message: The Application Default Credentials are not available. They are available if running in Google Compute Engine. Otherwise, the environment variable GOOGLE_APPLICATION_CREDENTIALS must be defined pointing to a file defining the credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information."

Login with developer credentials:
[,bash]
----
gcloud auth application-default login
----
====

== Traces

Open the https://console.cloud.google.com/traces[GCP Tracing Console].

image::gcp-tracing/trace-list-graph.png[]

{empty} +

> It does take a few seconds for the traces to show up in the console.

Reviewer: ```suggestion
If your traces aren't displayed yet, give it a moment. It takes a few seconds for the traces to show up in the console.
```

=== Get Item Counts

[source, bash]
----
curl http://localhost:8080/store/inventory/laptop
----
{empty} +

image::gcp-tracing/inventory-item.png[]

{empty} +

Each span is represented by a blue bar.

=== Order Item

[source, bash]
----
curl -X "POST" "http://localhost:8080/store/order" \
     -H 'Content-Type: application/json; charset=utf-8' \
     -d $'{"item":"laptop", "count":5}'
----

{empty} +

image::gcp-tracing/order-item.png[]

{empty} +

> Selecting a different span will show you the labels (a.k.a attributes/tags) and other details of the span.

Reviewer: ```suggestion
Selecting a different span will show you the labels (a.k.a. attributes/tags) and other details of the span.
```

=== Get Inventory

[source, bash]
----
curl http://localhost:8080/store/inventory
----

{empty} +

image::gcp-tracing/inventory-all.png[]

{empty} +

> Looking at the trace, retrieving the items sequentially might not be the best design choice.

Reviewer: ```suggestion
Looking at the trace, we can conclude that retrieving the items sequentially might not be the best design choice.
```
